feat(ci): Join FE & BE workflows

armenzg opened the pull request:
This adds two new checks (Frontend and Backend) that will be made required checks in the future.

This PR also moves all frontend jobs into the `frontend.yml` file and the backend jobs into `backend.yml`. PG14 jobs will be followed up in the future.

This is preparatory work to reduce the number of checks required to merge to `master`.

armenzg commented on .github/workflows/backend.yml:
```diff
@@ -0,0 +1,44 @@
+name: Backend
```
This workflow right now is just a placeholder for a following PR.

armenzg commented on .github/workflows/frontend.yml:
```diff
@@ -1,15 +1,27 @@
+          filters: .github/file-filters.yml
+
   typescript-and-lint:
+    needs: files-changed
```
Once these FE jobs are not required we will be able to add `if: ${{ needs.changes.outputs.frontend == 'true' }}` and remove the in-steps conditionals.

armenzg commented:
@billyvg what's this warning about? https://github.com/getsentry/sentry/runs/5887772069?check_suite_focus=true#step:8:72

billyvg commented:
@armenzg it tries to download an artifact from the same workflow but on master branch, but since it's a new workflow it doesn't exist

armenzg commented:
> @armenzg it tries to download an artifact from the same workflow but on master branch, but since it's a new workflow it doesn't exist

Thanks for confirming. It should auto-fix once this merges.

armenzg commented on .github/file-filters.yml:
```diff
@@ -56,15 +56,25 @@ backend: &backend
+backend_any_type: &backend_any_type
+  - *backend
+  - *api_docs
+  - *plugins
```
In the backend.yml we have jobs that check on either one of these. In order to handle the fake workflow we use this.

armenzg commented on .github/file-filters.yml:
```diff
@@ -56,15 +56,25 @@ backend: &backend
-plugins:
+acceptance: &acceptance
+  - *backend
+  - *frontend
```
This is used for the fake job on acceptance.yml.

armenzg commented on .github/file-filters.yml:
```diff
@@ -56,15 +56,25 @@ backend: &backend
+  - *backend
+  - *frontend
+
+plugins: &plugins
```
This enables me to use `*plugins` down below.

armenzg commented on .github/workflows/acceptance.yml:
```diff
@@ -14,6 +14,19 @@ on:
+        id: changes
+        with:
+          token: ${{ github.token }}
+          filters: .github/file-filters.yml
```
Now, only one job will make these API calls, thus, hopefully reducing the chances of `ERROR: Bad credentials`:
![image](https://user-images.githubusercontent.com/44410/162493296-3cc0920c-347f-4eb8-91fc-140a5dbb6c8f.png)

armenzg commented on .github/workflows/acceptance.yml:
```diff
@@ -212,17 +225,10 @@ jobs:
         uses: ./.github/actions/artifacts
-        if: ${{ steps.changes.outputs.backend == 'true' }}
+        # In order to determine the flag to use in codecov
+        if: ${{ needs.changes.outputs.backend == 'true' }}
```
s/steps/needs/

armenzg commented on .github/workflows/acceptance.yml:
```diff
@@ -314,3 +313,26 @@ jobs:
+  # This check is the only required Github check
+  acceptance-required-check:
+    needs: [visual-diff]
+    if: ${{ needs.changes.outputs.acceptance == 'true' }}
```
This if will eventually move up the dependency tree.

armenzg commented on .github/workflows/backend.yml:
```diff
@@ -0,0 +1,445 @@
+
+  api-docs:
+    needs: files-changed
+    name: api docs test
```
I have copy/pasted all the workflows I deleted. I added the line `needs: files-changed` and the key for each job (e.g. `test:` to `api-docs:`).

I also changed the conditional steps from: `steps` to `needs`. See below.

armenzg commented on .github/workflows/backend.yml:
```diff
@@ -0,0 +1,445 @@
+      - uses: actions/checkout@v2
+
+      - uses: volta-cli/action@v1
+        if: needs.changes.outputs.api_docs == 'true'
```
All the if steps have had this substitution: s/steps/needs/

armenzg commented on .github/workflows/backend.yml:
```diff
@@ -0,0 +1,445 @@
+        run: |
+          make setup-git
+
+      - uses: getsentry/paths-filter@v2
```
This is the only `getsentry/paths-filter@v2` that I have not removed since I do not want to mess it up.

armenzg commented on .github/workflows/backend.yml:
```diff
@@ -0,0 +1,445 @@
+        symbolicator,
+        typing,
+      ]
+    if: ${{ needs.changes.outputs.backend == 'true' }}
```
Eventually this condition will move to each one of the jobs and will be removed from here.

armenzg commented on .github/workflows/backend.yml:
```diff
@@ -0,0 +1,445 @@
+        snuba,
+        symbolicator,
+        typing,
+      ]
```
These are all the jobs defined above. All of them initially depend on `files-changed`.

billyvg replied:
Can you confirm that if a single `needs` job fails, this job will also fail?

Another thing that would be nice to confirm is the behavior of 1) if we have 1 job skipped, rest success and 2) if all jobs are skipped

armenzg replied:
Unless all of them pass, the required job will stay as skipped.

Over [here](https://github.com/getsentry/sentry/pull/32689#discussion_r832347061) shows that 1 check failed and it stayed skipped.

I had to re-run the failed check:
![image](https://user-images.githubusercontent.com/44410/162518448-90155f9c-2d31-4766-b9ae-281036b93e2e.png)

It even mentioned that "check to trigger VS" could get executed by re-running the single failure:
![image](https://user-images.githubusercontent.com/44410/162518521-b003ba58-5d70-42df-8242-00f4ae0e0417.png)

I can test your nice to confirm.

armenzg commented on .github/workflows/acceptance.yml:
```diff
@@ -314,3 +313,26 @@ jobs:
+  acceptance-required-check:
+    needs: [visual-diff]
+    if: ${{ needs.changes.outputs.acceptance == 'true' }}
+    name: Acceptance
```
<img width="299" alt="image" src="https://user-images.githubusercontent.com/44410/162495008-2a9a0a28-6951-4674-ae20-b8e6f728f263.png">
<img width="363" alt="image" src="https://user-images.githubusercontent.com/44410/162495341-37730c5c-6af2-4d34-b540-4570444d23c6.png">

armenzg commented on .github/workflows/backend.yml:
```diff
@@ -0,0 +1,445 @@
+        typing,
+      ]
+    if: ${{ needs.changes.outputs.backend == 'true' }}
+    name: Backend
```
<img width="328" alt="image" src="https://user-images.githubusercontent.com/44410/162495147-e387b60d-5b06-47c6-9799-4fdad0cd157a.png">

armenzg commented on .github/workflows/frontend.yml:
```diff
@@ -159,3 +172,26 @@ jobs:
+  frontend-required-check:
+    needs: [typescript-and-lint, webpack]
+    if: ${{ needs.changes.outputs.frontend == 'true' }}
+    name: Frontend
```
<img width="274" alt="image" src="https://user-images.githubusercontent.com/44410/162495205-cd4d0503-1bf3-47b1-a87f-12b10e1aa296.png">
<img width="325" alt="image" src="https://user-images.githubusercontent.com/44410/162495237-0d1001dd-3fff-4285-8c35-fb763011ccae.png">

armenzg reviewed (commented):
Notice that `backend.yml` shows as collapsed when you look at the diff.

billyvg commented on .github/workflows/acceptance.yml:
```diff
@@ -212,17 +225,10 @@ jobs:
       - name: Handle artifacts
         uses: ./.github/actions/artifacts
-        if: ${{ steps.changes.outputs.backend == 'true' }}
+        # In order to determine the flag to use in codecov
```
This doesn't determine the flag, it just completely skips uploading anything if it's not backend.

billyvg commented on .github/workflows/acceptance.yml:
```diff
@@ -314,3 +313,26 @@ jobs:
+  # This check runs once all dependant jobs have passed
+  # It symbolizes that all required Acceptance checks have succesfully passed
+  # This check is the only required Github check
+  acceptance-required-check:
+    needs: [visual-diff]
```
I don't think this will work that well. You'll still need Visual Snaphots as a required job because it can still fail (e.g. snapshot changes), but `visual-diff` will run fine because it triggers the diffing.

My line of thinking to have a separate Acceptance test "check" was in case people don't associate Visual Snapshots w/ acceptance tests. e.g. they scan only for "required" checks that have failed, but only see "Visual Snapshots" -- but this may not be the case anyway since people have asked about failing non-required checks in the past (e.g. during py3 migration).

I say try only having Visual Snapshots required like you initially suggested and go from there.

billyvg commented on .github/workflows/backend.yml:
```diff
@@ -0,0 +1,445 @@
+        run: |
+          yarn add ts-node && make test-api-docs
+
+  backend:
```
should we update this to `backend-test`?

billyvg commented on .github/workflows/backend.yml:
```diff
@@ -0,0 +1,445 @@
+  # In order to do this, we need to pretend that the real check run but hijacking the name
+  # This is the recommended approach by Github:
+  # https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/troubleshooting-required-status-checks#handling-skipped-but-required-checks
+  fake-backend-required-check:
+    needs: files-changed
+    if: ${{ needs.changes.outputs.backend_any_type != 'true' }}
+    name: Backend
+    runs-on: ubuntu-20.04
+    steps:
+      - run: 'echo "The changed files do not require running the checks in this workflow." '
```
This is redundant because we now only have a single job (`backend-required-check`), this hack was only necessary when we had matrices and we had to simulate all of the matrix names.

billyvg reviewed (changes requested):
There are a few issues atm:

* the `needs` logic is not working
* skipped jobs *satisfy* github required checks, so you can apply the if conditions to the jobs itself right now and it should work fine
* I'm unsure of behavior of skipped jobs and `needs`, I assume it also assumes it as a successful result, but I'm not certain.
* the fake jobs are now unnecessary because skipped jobs are considered "successful"

armenzg commented:
> There are a few issues atm:
> 
>     * the `needs` logic is not working

Fixed. I think.

>     * skipped jobs _satisfy_ github required checks, so you can apply the if conditions to the jobs itself right now and it should work fine
> 

I don't think that skipped jobs satisfy the requirement. This shows that I have 9 expected checks that have not started running. Those are the skipped checks from backend.yml
<img width="362" alt="image" src="https://user-images.githubusercontent.com/44410/162756789-43dec545-0222-4bf4-997d-e8c37c8a3a53.png">
![image](https://user-images.githubusercontent.com/44410/162756983-66f62912-7f98-4f03-85ad-e2f1755b5e22.png)
![image](https://user-images.githubusercontent.com/44410/162757018-c901a9ef-f00d-44c6-a329-51ed5fd99fad.png)



>     * I'm unsure of behavior of skipped jobs and `needs`, I assume it also assumes it as a successful result, but I'm not certain.

That's correct. In order to allow that last job to execute regardless of passing you would need to use `always()`.

1 job fails; the required check is marked skipped:
<img width="666" alt="image" src="https://user-images.githubusercontent.com/44410/162750043-050a0afe-1463-4311-bb7b-05cc4b203100.png">

>     * the fake jobs are now unnecessary because skipped jobs are considered "successful"

I've removed them but based on my investigation about skipped checks we will have to add them back.

> Can you confirm that if a single needs job fails, this job will also fail?

> Another thing that would be nice to confirm is the behavior of 1) if we have 1 job skipped, rest success and 2) if all jobs are skipped

One job skipped; the last one is also skipped:
<img width="729" alt="image" src="https://user-images.githubusercontent.com/44410/162753475-e8c7acf8-130d-485e-94d4-db4198b528d2.png">

All of them skipped, the last one is also skipped:
<img width="542" alt="image" src="https://user-images.githubusercontent.com/44410/162751341-9eaf03c5-60d4-4a8c-97c3-5749038809c7.png">

armenzg commented:
No expected check for webpack but only the skipped one:
<img width="563" alt="image" src="https://user-images.githubusercontent.com/44410/162761194-b99a792b-ceff-4410-adb0-f21e488bb4a4.png">
The only other expected check is "Visual Snapshot":
<img width="360" alt="image" src="https://user-images.githubusercontent.com/44410/162761271-df827bfb-ab39-4d7e-8cbb-3f274ebfc0e4.png">
<img width="519" alt="image" src="https://user-images.githubusercontent.com/44410/162761316-417a0956-2600-46b5-a51b-8fd2576d8a73.png">

armenzg commented:
@billyvg Very grateful for all that you caught on your review!

I've moved the per step conditions to the top of each job.

This is what [a PR](https://github.com/getsentry/sentry/pull/33501) based on this that does *not* match FE or BE file matches looks like (23 skipped checks! - the pending checks are Vercel):
<img width="363" alt="image" src="https://user-images.githubusercontent.com/44410/162771777-688654a5-3478-4e96-beea-9b49fb0738c2.png">

The number of skipped checks will get better once I merge PG14 jobs back to where they belong (on a future PR).

billyvg commented:
> I've moved the per step conditions to the top of each job.

@armenzg Don't we still need the per step conditions until we're ready to switch the required checks? Otherwise the skipped jobs will fail required checks due to the lack of matrix expansion. Your original thinking of this being a two step process where PR #1 is where we skip per step, but the important bit is making sure the Frontend/Backend checks are being correctly created. Then after a certain period of time where we're confident most PRs are going to be up-to-date with that change, we can land PR #2 to skip at the job level and switch the Required checks.

That being said, this PR LGTM as PR #2!

armenzg commented:
> > I've moved the per step conditions to the top of each job.
> 
> @armenzg Don't we still need the per step conditions until we're ready to switch the required checks? Otherwise the skipped jobs will fail required checks due to the lack of matrix expansion. Your original thinking of this being a two step process where PR #1 is where we skip per step, but the important bit is making sure the Frontend/Backend checks are being correctly created. Then after a certain period of time where we're confident most PRs are going to be up-to-date with that change, we can land PR #2 to skip at the job level and switch the Required checks.
> 
> That being said, this PR LGTM as PR #2!

That's changed thanks to your discovery that Skipped checks count toward satisfying the required checks criteria.
We don't need to do this in two steps anymore.

I also got the impression that you asked me to do it based on this comment :)

> skipped jobs satisfy github required checks, so you can apply the if conditions to the jobs itself right now and it should work fine

We will have to wait to change the list of required checks via the UI (or ask everyone to rebase/merge).

The [other PR](https://github.com/getsentry/sentry/pull/33501) proves that we skip the jobs and we still satisfy Github's requirements:
<img width="746" alt="image" src="https://user-images.githubusercontent.com/44410/162965743-e950a42a-1a14-439f-9ddf-bceef895c35f.png">
<img width="383" alt="image" src="https://user-images.githubusercontent.com/44410/162966103-75d49146-7e5e-4521-ab12-f5df0ea0d4ce.png">

billyvg commented:
We discussed this offline but for posterity: when you skip a [required] job that has a matrix, GH required checks will not be met. This is because GH

1) appends the matrix values of a job to the check name
2) uses the check name (with matrix values) when configuring required checks 
3) skipped jobs will never run the matrix, so check names will only have the job name

Thus the required check(s) will never be met.

See https://github.com/github/feedback/discussions/9141

billyvg reviewed (approved):
🚢
